### PR TITLE
Add NeighborValueDot

### LINKED ISSUE #12372: Add coupledNeighborValueDot to NeighborCoupleable
## Rationale

Some physics requires a residual on the time derivative of a variable along an interface. 

## Description

For NeighborCoupleable
- Added a virtual function ('coupledNeighborValueDot') returning the time derivative of neighbor variable (_u_dot_neighbor)
- Add a virtual function ('coupledNeighborValueDotDu') returning the derivative of _u_dot_neighbor  w.r.t. _u_neighbor (_du_dot_neighbor_du)

For Coupleable fixed 'coupledDotDu' (ref #12380) making sure it returns the appropriate value for neighbors

For interfaceKernel:
added time derivative quantities for the master and the slave side. Also some minor improvement for code readability. 

## Impact

No negative impact, simply a new feature

### PR DESCRIPTION
**Pull request by arovinelli**
close #12372, #12380 

Added a virtual function returning the time derivative of a variable.

### CODE REVIEW

**Review by lindsayad — CHANGES_REQUESTED** [review]
I'm glad that this involved pretty little work! Could you provide a test for this?

**Comment by arovinelli** [comment]
Dear Alex, I'm trying to make a test in an interface kernel but it's becoming more involved than I though.
I basically added time dependent variables to the IK and a set a residual of the type
R = test[i]*K*(u_dot_master - u_dot_slave), set the Jacobian. However it does not converge as it's supposed to be with newton.

File: test/src/interfacekernels/PenaltyInterfaceDiffusionDot.C [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    case Moose::Neighbor:
+      r = _test_neighbor[_i][_qp] * -_penalty * (_u_dot[_qp] - _neighbor_value_dot[_qp]);
+      break;
+  }
+
+  return r;
+}
+
+Real
+PenaltyInterfaceDiffusionDot::computeQpJacobian(Moose::DGJacobianType type)
```
Comment (arovinelli) [inline test/src/interfacekernels/PenaltyInterfaceDiffusionDot.C]: I'm can't figure out what's wrong with the Jacobian.

Comment (YaqiWang) [reply test/src/interfacekernels/PenaltyInterfaceDiffusionDot.C]: Did you try PJFNK solve_type?

Comment (arovinelli) [reply test/src/interfacekernels/PenaltyInterfaceDiffusionDot.C]: @YaqiWang, yes I tried and it works. However, I always try to use newton to make sure I have the right jacobian. The thing is newton works with FDP but not with the supplied, ideally exact, jacobian. So it should be messed-up somehow

Comment (arovinelli) [reply test/src/interfacekernels/PenaltyInterfaceDiffusionDot.C]: Ok, I figured it out. I t was my mistake in defying _u_dot(_var.uDot()),

**Comment by arovinelli** [comment]
This is ready for review!

**Review by lindsayad — CHANGES_REQUESTED** [review]
@arovinelli Have you ever used Method of Manufactured Solutions (MMS)? I'm a little worried that we would not demonstrate optimal convergence with respect to time for explicit or mixed implicit/explicit time integration schemes with the current set-up. We have `TimeKernel` as a derived class of `Kernel` for this purpose. We may need an `InterfaceTimeKernel`. If you're familiar with MMS, I'd invite you to try some convergence studies. If you're not, then I will. This originated as a simple bug fix for you so I don't want to ask you to do a ton of work here.

**Comment by arovinelli** [comment]
@lindsayad I've never used MMS and I'm not familiar with it. I suppose you want to decouple spatial and time dependent residual such that you can scale them appropriately. Adding an 'InterfaceTimeKernel' does also make sense to make things more modular.

- The IK I implemented (InterfacePenaltyDot) is just to show that the new additions in NeighborCouplable work correctly.
- The addition of the time derivative to are of a more general use. I could look at the 'TimeKernel' and translate those ideas into an 'InterfaceTimeKernel'. This hsould not be too painful (subclass IK and add the time dependent varaibles). 

Let me know how you want to proceed with this.

**Comment by lindsayad** [comment]
That sounds great!

**Comment by arovinelli** [comment]
@lindsayad I added an 'InterfaceTimeKernel' and derived the test from it.
You may want to take a look at it

**Comment by lindsayad** [comment]
Unity build failing...one potential reason is missing includes in your source file

**Comment by arovinelli** [comment]
> Unity build failing...one potential reason is missing includes in your source file

@lindsayad It should be fixed... However, is there a way to use Non Unity Build locally (so that I can easily check before pushing)?

**Comment by lindsayad** [comment]
`export MOOSE_UNITY=false` in your environment

**Review by lindsayad — CHANGES_REQUESTED** [review]
This looks really good. Just a couple of small changes

File: framework/src/interfacekernels/InterfaceKernel.C [hunk 2]
```diff
@@ -101,14 +101,18 @@ InterfaceKernel::InterfaceKernel(const InputParameters & parameters)
...
     _phi(_assembly.phiFace(_var)),
     _grad_phi(_assembly.gradPhiFace(_var)),
     _test(_var.phiFace()),
     _grad_test(_var.gradPhiFace()),
     _normals(_var.normals()),
     _neighbor_var(*getVar("neighbor_var", 0)),
-    _neighbor_value(_neighbor_var.slnNeighbor()),
+    _neighbor_value(coupledNeighborValue("neighbor_var")),
+    // _neighbor_value_dot(_is_transient ? coupledNeighborValueDot("neighbor_var") : _zero),
+    // _dneighbor_value_dot_du(_is_transient ? coupledNeighborValueDotDu("neighbor_var") : _zero),
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply framework/src/interfacekernels/InterfaceKernel.C]: oops I forgot about that!

File: framework/src/interfacekernels/InterfaceTimeKernel.C [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "InterfaceTimeKernel.h"
+
+template <>
+InputParameters
+validParams<InterfaceTimeKernel>()
+{
+  InputParameters params = validParams<InterfaceKernel>();
+
+  return params;
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply framework/src/interfacekernels/InterfaceTimeKernel.C]: Done!

[the remark this answers is not in the document]
Comment (arovinelli) [reply framework/src/interfacekernels/InterfaceTimeKernel.C]: This two lines make my test fail...
```
interfacekernels/2d_interface.test_dot: Working Directory: /opt/civet/build_0/moose/test/tests/interfacekernels/2d_interface
interfacekernels/2d_interface.test_dot: Running command: /opt/civet/build_0/moose/test/moose_test-opt -i coupled_value_coupled_flux_dot.i --error --error-unused --error-override --no-gdb-backtrace
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: *** ERROR ***
interfacekernels/2d_interface.test_dot: Invalid option "time" in MultiMooseEnum.  Valid options (not case-sensitive) are "".
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: Stack frames: 21
interfacekernels/2d_interface.test_dot: 0: libMesh::print_trace(std::ostream&)
interfacekernels/2d_interface.test_dot: 1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
interfacekernels/2d_interface.test_dot: 2: MultiMooseEnum::operator=(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
interfacekernels/2d_interface.test_dot: 3: InputParameters validParams<InterfaceTimeKernel>()
interfacekernels/2d_interface.test_dot: 4: InputParameters validParams<PenaltyInterfaceDiffusionDot>()
interfacekernels/2d_interface.test_dot: 5: Factory::getValidParams(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
interfacekernels/2d_interface.test_dot: 6: MooseObjectAction::MooseObjectAction(InputParameters)
interfacekernels/2d_interface.test_dot: 7: AddInterfaceKernelAction::AddInterfaceKernelAction(InputParameters)
interfacekernels/2d_interface.test_dot: 8: std::shared_ptr<Action> buildAct<AddInterfaceKernelAction>(InputParameters const&)
interfacekernels/2d_interface.test_dot: 9: ActionFactory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters)
interfacekernels/2d_interface.test_dot: 10: Parser::walkRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, hit::Node*)
interfacekernels/2d_interface.test_dot: 11: Parser::walk(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, hit::Node*)
interfacekernels/2d_interface.test_dot: 12: hit::Node::walk(hit::Walker*, hit::NodeType)
interfacekernels/2d_interface.test_dot: 13: hit::Node::walk(hit::Walker*, hit::NodeType)
interfacekernels/2d_interface.test_dot: 14: hit::Node::walk(hit::Walker*, hit::NodeType)
interfacekernels/2d_interface.test_dot: 15: Parser::parse(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
interfacekernels/2d_interface.test_dot: 16: MooseApp::setupOptions()
interfacekernels/2d_interface.test_dot: 17: MooseApp::run()
interfacekernels/2d_interface.test_dot: 18: /opt/civet/build_0/moose/test/moose_test-opt() [0x4019bd]
interfacekernels/2d_interface.test_dot: 19: __libc_start_main
interfacekernels/2d_interface.test_dot: 20: /opt/civet/build_0/moose/test/moose_test-opt() [0x401819]
interfacekernels/2d_interface.test_dot: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
interfacekernels/2d_interface.test_dot: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
interfacekernels/2d_interface.test_dot: :
interfacekernels/2d_interface.test_dot: system msg for write_line failure : Bad file descriptor
interfacekernels/2d_interface.test_dot: 
interfacekernels/2d_interface.test_dot: ################################################################################
interfacekernels/2d_interface.test_dot: Tester failed, reason: ERRMSG
interfacekernels/2d_interface.test_dot: 
```

**Comment by lindsayad** [comment]
Uh oh, I forgot that we haven't added the `TaggingInterface` to `InterfaceKernel` yet. [This PR](https://github.com/idaholab/moose/pull/12396) does it, so this PR should be able to go in after that one.

**Comment by arovinelli** [comment]
OK, so we should rerun the tests once #12396 get merged. Any timeline for it? I'm starting to have too many open PR hanging out and keeping everything update is a nightmare :(

**Comment by lindsayad** [comment]
@arovinelli two days was a little ambitious, but #12396 is now merged

**Comment by arovinelli** [comment]
@lindsayad great! Thanks for letting me know! Do I have to rebase or you can run tests??

**Comment by arovinelli** [comment]
@lindsayad dear Alex, please take a look at the PR and let me know if It requires other changes/improvements, including variables naming. Thanks

**Comment by lindsayad** [comment]
So there will be a temporary hold-up because we actually had to revert @fdkong's tagging PR. This should be fixed soon and then this can be merged

**Review by lindsayad — APPROVED** [review]

**Pull request merged**